job/job.go

```diff
@@ -0,0 +1,406 @@
+// Copyright 2023 tsuru authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package job
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/adhocore/gronx"
+	"github.com/globalsign/mgo"
+	"github.com/imdario/mergo"
+	"github.com/pkg/errors"
+	"github.com/tsuru/tsuru/action"
+	"github.com/tsuru/tsuru/auth"
+	"github.com/tsuru/tsuru/db"
+	tsuruErrors "github.com/tsuru/tsuru/errors"
+	"github.com/tsuru/tsuru/provision"
+	"github.com/tsuru/tsuru/provision/pool"
+	"github.com/tsuru/tsuru/servicemanager"
+	"github.com/tsuru/tsuru/set"
+	appTypes "github.com/tsuru/tsuru/types/app"
+	jobTypes "github.com/tsuru/tsuru/types/job"
+	"gopkg.in/mgo.v2/bson"
+)
+
+// Job is another main type in tsuru as of version *insert current version*.
+// A job currently represent a Kubernetes Job object or a Cronjob object
+// this struct is composited of a TsuruJob and all the metadata tsuru natively uses
+// as well as job specific definitions such as schedule and container image and commands
+type Job struct {
```

Comment:
Three considerations about this struct:

1. I'd take out this embedding struct. It's more clear merging the `TsuruJob` with `Job` and move the definitions away to other struct. Example:
```golang
type Job struct {
	Name  string     `json:"name"`
	Teams []string  `json:"teams"`
// ...
	Spec    JobSpec  `json:"jobSpec,omitempty"`
}

type JobSpec struct {
	Container jobTypes.ContainerInfo `json:"container"`
}
```

2. Consider moving it for `types` package (e.g. `types/jobs/`). Just to follow the flow of other types.
3. Remove protobuf tags. Probably you copied from Kubernetes types, but here they're not necessary.

Reply:
> Consider moving it for types package (e.g. types/jobs/). Just to follow the flow of other types.
I copied the way we did with the _app_ resource, we only declare it's interface in the _types_ package, but the struct definition we keep it in the _app_ package

Reply:
@nettoclaudio could you take another look? I refactored but kept the job struct inside the job package. If you still think it would be better to move it to _types/job_ please say so.